qiita_db/processing_job.py

```diff
@@ -219,6 +221,60 @@ def status(self, value):
             qdb.sql_connection.TRN.add(sql, [new_status, self.id])
             qdb.sql_connection.TRN.execute()
 
+    def complete(self, success, artifacts_data=None, error=None):
+        """Completes the job, either with a success or error status
+
+        Parameters
+        ----------
+        success : bool
+            Whether the job has completed successfully or not
+        artifacts_data : dict of dicts
+            The generated artifact information, keyed by output name.
+            The format of each of the internal dictionaries must be
+            {'filepaths': list of (str, str), 'artifact_type': str}
+            where `filepaths` contains the list of filepaths and filepath types
+            for the artifact and `artifact_type` the type of the artifact
+        error : str, optional
+            If the job was not successful, the error message
+
+        Raises
+        ------
+        qiita_db.exceptions.QiitaDBOperationNotPermittedError
+            If the job is not in running state
+        """
+        with qdb.sql_connection.TRN:
+            if self.status != 'running':
+                raise qdb.exceptions.QiitaDBOperationNotPermittedError(
+                    "Can't complete job: not in a running state")
+            if success:
+                if artifacts_data:
+                    artifact_ids = []
+                    for out_name, a_data in viewitems(artifacts_data):
+                        filepaths = a_data['filepaths']
+                        atype = a_data['artifact_type']
+                        parents = self.input_artifacts
+                        params = self.parameters
+                        a = qdb.artifact.Artifact.create(
+                            filepaths, atype, parents=parents,
+                            processing_parameters=params)
+                        cmd_out_id = qdb.util.convert_to_id(
+                            out_name, "command_output", "name")
+                        artifact_ids.append((cmd_out_id, a.id))
+                    if artifact_ids:
```

Comment:
Currently there is no command that doesn't generate an artifact. I'm preparing the system for such cases. One example may be EBI submission - they don't generate new artifacts. Since EBI submission is not a plugin command, I can't test.